addons/website_sale_delivery/models/sale_order.py

```diff
@@ -38,6 +38,7 @@ def _check_carrier_quotation(self, force_carrier_id=None):
             return True
         else:
             self = self.with_company(self.company_id)
+            self.validate_taxes_on_sales_order()
```

Comment:
> Shouldn't we call it before the `if self.only_services:´ instead ?

I don't know if it should be before `if self.only_services:` as in this condition, it will just remove the delivery line.
But it could be without any problem.

> Also, if we call validate_taxes_on_sales_order in _check_carrier_quotation, can't we remove the call in the _update_website_sale_delivery_return override of website_sale_taxcloud_delivery ?
Does it make sense to call it once before the carrier update and once after as well ?

It can be removed, but another call of `self.validate_taxes_on_sales_order()` is required at the end of the `_check_carrier_quotation` method to compute the tax on the delivery when we have several ones with different prices and we switch between them on the payment page.

Reply:
Mhm, with my non existent knowledge of avatax & taxcloud, I dunno for certainty exactly in which cases the call to the taxes validation is required or not :D.
Furthermore, there has been a lot of changes in the code between 15.0 and master.

After more investigation, AFAICS, this is already a mess :pensive:.

Right now, in 15, taxcloud taxes are recomputed:
* `website_sale_account_taxcloud`: on payment page loading (`_get_shop_payment_errors` override), too late to be considered before the call to `_check_carrier_quotation` in the `shop_payment` override.
* `website_sale_taxcloud_delivery`: after carrier update (`_update_website_sale_delivery_return` override)
* `website_sale_coupon_delivery` only: after the carrier & promotions update (update_eshop_carrier override)
* `website_sale_account_taxcloud`: on transaction step (shop_payment_transaction override)

IIUC, the issue comes from the fact that, when setting the default carrier (on /shop/payment first loading), the taxes have not been properly computed, and the associated carrier price is computed on wrong amounts.

I've thought about it, and adding another recomputation of the taxes in each call to `_check_carrier_quotation` is probably overkill.  Furthermore, it does not gracefully handle failures of `validate_taxes_on_sales_order` (which is correctly handled in `_get_shop_payment_errors` override), which could lead to a bad user experience if such an error is not catched nor displayed cleanly.

Can't we solve the problem we're facing by simply adding a call to `validate_taxes_on_sales_order` in `/shop/confirm_order`, after `order.onchange_partner_shipping_id()` and `order._compute_tax_id()` (which ideally shouldn't do anything for taxcloud fpos, but it seems it's not handled, unlike avatax) and in case of failure, simply skip without displaying anything (/shop/payment will handle it already) ?
That way, the amounts will be the right ones before even landing on the `/shop/payment` page :thinking:

Reply:
> Can't we solve the problem we're facing by simply adding a call to validate_taxes_on_sales_order in `/shop/confirm_order`?

Yes, we can. I've updated both fixes.